hpx/runtime/actions/transfer_base_action.hpp

```diff
@@ -19,24 +19,124 @@
 #include <hpx/runtime/serialization/base_object.hpp>
 #include <hpx/runtime/serialization/input_archive.hpp>
 #include <hpx/runtime/serialization/output_archive.hpp>
+#include <hpx/runtime/serialization/unique_ptr.hpp>
 #include <hpx/traits/action_does_termination_detection.hpp>
 #include <hpx/traits/action_message_handler.hpp>
 #include <hpx/traits/action_was_object_migrated.hpp>
 #include <hpx/traits/action_priority.hpp>
 #include <hpx/traits/action_schedule_thread.hpp>
 #include <hpx/traits/action_serialization_filter.hpp>
 #include <hpx/traits/action_stacksize.hpp>
+#include <hpx/util/assert.hpp>
 #include <hpx/util/get_and_reset_value.hpp>
 #include <hpx/util/serialize_exception.hpp>
+#include <hpx/util/tuple.hpp>
 
 #include <boost/atomic.hpp>
 
 #include <cstddef>
 #include <cstdint>
+#include <memory>
+#include <type_traits>
 #include <utility>
 
 namespace hpx { namespace actions
 {
+    ///////////////////////////////////////////////////////////////////////////
+    // If one or more arguments of the action are non-default-constructible,
+    // the transfer_action does not store the argument tuple directly but a
+    // unique_ptr to the tuple instead.
```

Comment:
I decided to risk adding one allocation for the case when the type is not default-constructible. This shouldn't happen more than in 1% of all use cases, thus the effect wouldn't be measurable. Let's go ahead with what's proposed and fix it if we come across a real show-stopper.